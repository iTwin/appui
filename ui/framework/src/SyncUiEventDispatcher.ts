/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import { UiEvent } from "@bentley/ui-core";

/** SyncUi Event arguments. Contains a set of lower case event Ids.
 */
export interface SyncUiEventArgs {
  eventIds: Set<string>;
}

/** SyncUi Event class.
 */
export class SyncUiEvent extends UiEvent<SyncUiEventArgs> { }

/** SyncUi Event Dispatcher class. This class is used to send eventIds to interested Ui components so the component can determine if it needs
 * to refresh its display by calling setState on itself.
 */
export class SyncUiEventDispatcher {
  private static _syncEventTimer: NodeJS.Timer | undefined;
  private static _eventIds: Set<string>;
  private static _eventIdAdded: boolean = false;
  private static _syncUiEvent: SyncUiEvent;
  private static _timeoutPeriod: number = 500;

  /** @hidden - used for testing only */
  public static setTimeoutPeriod(period: number): void {
    SyncUiEventDispatcher._timeoutPeriod = period;
  }

  /** Return SyncUiEvent so callers can register an event callback. */
  public static get syncEventIds(): Set<string> {
    if (!SyncUiEventDispatcher._eventIds)
      SyncUiEventDispatcher._eventIds = new Set<string>();

    return SyncUiEventDispatcher._eventIds;
  }

  /** Return SyncUiEvent so callers can register an event callback. */
  public static get onSyncUiEvent(): SyncUiEvent {
    if (!SyncUiEventDispatcher._syncUiEvent)
      SyncUiEventDispatcher._syncUiEvent = new SyncUiEvent();

    return SyncUiEventDispatcher._syncUiEvent;
  }

  /** Immediately trigger sync event processing. */
  public static dispatchImmediateSyncUiEvent(eventId: string): void {
    const eventIds = new Set<string>();
    eventIds.add(eventId.toLowerCase());
    SyncUiEventDispatcher.onSyncUiEvent.emit({ eventIds });
  }

  /** Save eventId in Set for processing. */
  public static dispatchSyncUiEvent(eventId: string): void {
    SyncUiEventDispatcher.syncEventIds.add(eventId.toLowerCase());
    SyncUiEventDispatcher._eventIdAdded = true;
    if (!SyncUiEventDispatcher._syncEventTimer) {  // if there is not a timer active, create one
      SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
    }
  }

  /** Save multiple eventIds in Set for processing. */
  public static dispatchSyncUiEvents(eventIds: string[]): void {
    eventIds.forEach((id) => SyncUiEventDispatcher.syncEventIds.add(id.toLowerCase()));
    if (!SyncUiEventDispatcher._syncEventTimer) {  // if there is not a timer active, create one
      SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
    } else {
      SyncUiEventDispatcher._eventIdAdded = true;
    }
  }

  /** Trigger registered event processing when timer has expired and no addition eventId are added. */
  private static checkForAdditionalIds() {
    if (!SyncUiEventDispatcher._eventIdAdded && SyncUiEventDispatcher._syncEventTimer) {
      if (SyncUiEventDispatcher._syncEventTimer) clearTimeout(SyncUiEventDispatcher._syncEventTimer);
      SyncUiEventDispatcher._syncEventTimer = undefined;
      SyncUiEventDispatcher._eventIdAdded = false;
      SyncUiEventDispatcher.onSyncUiEvent.emit({ eventIds: SyncUiEventDispatcher.syncEventIds });
      return;
    }

    if (SyncUiEventDispatcher._syncEventTimer) clearTimeout(SyncUiEventDispatcher._syncEventTimer);
    SyncUiEventDispatcher._eventIdAdded = false;
    SyncUiEventDispatcher._syncEventTimer = setTimeout(SyncUiEventDispatcher.checkForAdditionalIds, SyncUiEventDispatcher._timeoutPeriod);
  }
}
