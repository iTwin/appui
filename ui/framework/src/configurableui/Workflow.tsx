/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Workflow */

import { UiEvent } from "@bentley/ui-core";

import { IconLabelProps } from "./IconLabelSupport";
import ItemDefBase from "./ItemDefBase";
import { ItemProps } from "./ItemProps";
import { Task, TaskManager } from "./Task";

// -----------------------------------------------------------------------------
//  WorkflowDef and WorkflowsDef
// -----------------------------------------------------------------------------

/** Properties for a Workflow.
 */
export interface WorkflowProps extends ItemProps {
  id: string;
  defaultTaskId: string;
  tasks: string[];
  isDefault?: boolean;
}

/** Properties for a TaskPicker.
 */
export interface TaskPickerProps extends IconLabelProps {
  classid: string;
}

/** Workflow Properties List definition.
 */
export interface WorkflowPropsList {
  defaultWorkflowId: string;
  taskPicker: TaskPickerProps;
  workflows: WorkflowProps[];
}

// -----------------------------------------------------------------------------
// Workflow class
// -----------------------------------------------------------------------------

/** Workflow class.
 */
export class Workflow extends ItemDefBase {
  /** Id of the Workflow */
  public workflowId: string;
  /** Default Task Id */
  public defaultTaskId: string;
  /** Active Task Id */
  public activeTaskId: string | null = null;
  /** Indicates whether this Workflow is the default */
  public isDefault: boolean;

  private _taskIds: string[];
  private _tasks: Map<string, Task> = new Map<string, Task>();

  constructor(workflowDef: WorkflowProps) {
    super(workflowDef);

    this.workflowId = workflowDef.id;
    this.defaultTaskId = workflowDef.defaultTaskId;
    this._taskIds = workflowDef.tasks;
    this.isDefault = (workflowDef.isDefault !== undefined) ? workflowDef.isDefault : true;

    this._taskIds.map((taskId: string, _index: number) => {
      const task = TaskManager.findTask(taskId);
      if (task)
        this._tasks.set(taskId, task);
    });
  }

  /** Gets the Id of the Workflow. */
  public get id(): string {
    return this.workflowId;
  }

  /** Override of the execute method. A Workflow's execute does nothing. */
  public async execute(): Promise<void> {
  }

  /** Override of the toolbarReactNode method. A Workflow provides no Toolbar React node. */
  public toolbarReactNode(_index: number): React.ReactNode {
    return null;
  }

  /** Gets the active Task. */
  public get activeTask(): Task | undefined {
    if (!this.activeTaskId)
      return undefined;

    return this._tasks.get(this.activeTaskId);
  }

  /** Gets a Task with a given Id.
   * @param taskId Id of the Task to get
   */
  public getTask(taskId: string): Task | undefined {
    return this._tasks.get(taskId);
  }

  /** Gets the last active Task. If no Task is active, it returns the default Task.
   */
  public get lastActiveTask(): Task | undefined {
    if (!this.activeTaskId) {
      return this.getTask(this.defaultTaskId);
    }

    return this.getTask(this.activeTaskId);
  }

  /** Determines if the Workflow is active.
   */
  public get isActive(): boolean {
    let isActive = false;

    if (WorkflowManager.activeWorkflow === this)
      isActive = true;

    return isActive;
  }

  /** Sets a Task as active.
   * @param task  The Task to set as active
   */
  public setActiveTask(task: Task) {
    this.activeTaskId = task.taskId;
    task.onActivated();
    WorkflowManager.onTaskActivatedEvent.emit({ task, taskId: task.id });
  }

  /** Gets an array of sorted Tasks in the Workflow. */
  public getSortedTasks(): Task[] {
    const sortedTasks = new Array<Task>();

    for (const key of this._tasks.keys()) {
      const task: Task | undefined = this._tasks.get(key);
      if (task && task.isVisible)
        sortedTasks.push(task);
    }

    sortedTasks.sort((a: Task, b: Task) => {
      return a.label.localeCompare(b.label);
    });

    return sortedTasks;
  }

}

/** Workflow Activated Event Args class.
 */
export interface WorkflowActivatedEventArgs {
  workflowId: string;
  workflow: Workflow;
}

/** Workflow Activated Event class.
 */
export class WorkflowActivatedEvent extends UiEvent<WorkflowActivatedEventArgs> { }

/** Task Activated Event Args class.
 */
export interface TaskActivatedEventArgs {
  taskId: string;
  task: Task;
}

/** Task Activated Event class.
 */
export class TaskActivatedEvent extends UiEvent<TaskActivatedEventArgs> { }

// -----------------------------------------------------------------------------
// WorkflowManager class
// -----------------------------------------------------------------------------

/** Workflow Manager class.
 */
export class WorkflowManager {
  private static _workflows: Map<string, Workflow> = new Map<string, Workflow>();
  private static _activeWorkflow: Workflow;
  private static _defaultWorkflowId: string;
  private static _taskPickerProps: TaskPickerProps;
  private static _workflowActivatedEvent: WorkflowActivatedEvent = new WorkflowActivatedEvent();
  private static _taskActivatedEvent: TaskActivatedEvent = new TaskActivatedEvent();

  public static get onWorkflowActivatedEvent(): WorkflowActivatedEvent { return this._workflowActivatedEvent; }
  public static get onTaskActivatedEvent(): TaskActivatedEvent { return this._taskActivatedEvent; }

  /** Loads one or more Workflows.
   * @param workflowPropsList  the list of Workflows to load
   */
  public static loadWorkflows(workflowPropsList: WorkflowPropsList) {
    this._defaultWorkflowId = workflowPropsList.defaultWorkflowId;
    this._taskPickerProps = workflowPropsList.taskPicker;
    WorkflowManager.loadWorkflowDefs(workflowPropsList.workflows);
  }

  private static loadWorkflowDefs(workflowDefs: WorkflowProps[]) {
    workflowDefs.map((workflowProps: WorkflowProps) => {
      WorkflowManager.loadWorkflow(workflowProps);
    });
  }

  /** Loads a Workflow.
   * @param workflowProps  Properties of the Workflow to load
   */
  public static loadWorkflow(workflowProps: WorkflowProps) {
    const workflow = new Workflow(workflowProps);
    WorkflowManager.addWorkflow(workflowProps.id, workflow);
  }

  /** Adds a Workflow.
   * @param workflowId  Id of the Workflow to add
   * @param workflow    The Workflow to add
   */
  public static addWorkflow(workflowId: string, workflow: Workflow) {
    this._workflows.set(workflowId, workflow);
  }

  /** Finds a Workflow with a given Id.
   * @param workflowId  Id of the Workflow to find
   * @returns The Workflow if found, or undefined if not found
   */
  public static findWorkflow(workflowId: string): Workflow | undefined {
    return this._workflows.get(workflowId);
  }

  /** Sets the active Workflow
   * @param workflow  The Workflow to set as active
   */
  public static setActiveWorkflow(workflow: Workflow): void {
    this._activeWorkflow = workflow;
    WorkflowManager.onWorkflowActivatedEvent.emit({ workflow, workflowId: workflow.id });
  }

  /** Sets the active Workflow and Task
   * @param workflow  The Workflow to set as active
   * @param task      The Task to set as active
   */
  public static async setActiveWorkflowAndTask(workflow: Workflow, task: Task): Promise<void> {
    if (!workflow.isActive)
      this.setActiveWorkflow(workflow);

    if (!task.isActive)
      await workflow.setActiveTask(task);
  }

  /** Gets the active Workflow */
  public static get activeWorkflow(): Workflow {
    return this._activeWorkflow;
  }

  /** Gets the Id of the default Workflow */
  public static get defaultWorkflowId(): string {
    return this._defaultWorkflowId;
  }

  /** Gets the Task Picker properties */
  public static get taskPickerProps(): TaskPickerProps {
    return this._taskPickerProps;
  }

  public static getSortedWorkflows(): Workflow[] {
    const sortedWorkflows = new Array<Workflow>();

    for (const key of this._workflows.keys()) {
      const workflow = this._workflows.get(key);
      if (workflow) {
        sortedWorkflows.push(workflow);
      }
    }

    sortedWorkflows.sort((a: Workflow, b: Workflow) => {
      return a.label.localeCompare(b.label);
    });

    return sortedWorkflows;
  }
}
