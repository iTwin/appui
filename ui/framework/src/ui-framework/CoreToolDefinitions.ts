/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

// cSpell:ignore configurableui
import {
  FitViewTool, FlyViewTool, IModelApp, PanViewTool, RotateViewTool, SelectionTool, ViewToggleCameraTool, WalkViewTool,
  WindowAreaTool, ZoomViewTool, ViewClipByPlaneTool,
  ViewClipDecorationProvider,
} from "@bentley/imodeljs-frontend";
import { ViewFlags } from "@bentley/imodeljs-common";
import { ToolItemDef } from "./shared/ToolItemDef";

/** Utility Class that provides definitions of tools provided by iModel.js core. These definitions can be used to populate the UI.
 * @public
 */
// istanbul ignore next
export class CoreTools {
  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: "icon-fit-to-view",
      label: () => FitViewTool.flyover,
      tooltip: () => FitViewTool.description,
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: "icon-window-area",
      label: () => WindowAreaTool.flyover,
      tooltip: () => WindowAreaTool.description,
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: "icon-zoom",
      label: () => ZoomViewTool.flyover,
      tooltip: () => ZoomViewTool.description,
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: "icon-hand-2",
      label: () => PanViewTool.flyover,
      tooltip: () => PanViewTool.description,
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: "icon-gyroscope",
      label: () => RotateViewTool.flyover,
      tooltip: () => RotateViewTool.description,
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: "icon-walk",
      label: () => WalkViewTool.flyover,
      tooltip: () => WalkViewTool.description,
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: "icon-cursor",
      label: () => SelectionTool.flyover,
      tooltip: () => SelectionTool.description,
      execute: () => { IModelApp.tools.run(SelectionTool.toolId); },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: "icon-camera",
      label: () => ViewToggleCameraTool.flyover,
      tooltip: () => ViewToggleCameraTool.description,
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: "icon-airplane",
      label: () => FlyViewTool.flyover,
      tooltip: () => FlyViewTool.description,
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  // note current ViewClipByPlaneTool is not automatically registered so the app must call ViewClipByPlaneTool.register();
  public static get sectionByPlaneCommand() {
    return new ToolItemDef({
      toolId: ViewClipByPlaneTool.toolId,
      iconSpec: "icon-plane",
      label: () => ViewClipByPlaneTool.flyover,
      tooltip: () => ViewClipByPlaneTool.description,
      execute: () => {
        const vp = IModelApp.viewManager.selectedView;
        if (!vp || !vp.view.is3d())
          return;

        // Turn on clip volume flag for section tools
        const viewFlags: ViewFlags = vp.view.viewFlags.clone();
        viewFlags.clipVolume = true;
        vp.viewFlags = viewFlags;

        IModelApp.tools.run(ViewClipByPlaneTool.toolId, ViewClipDecorationProvider.create());
      },
    });
  }

}
