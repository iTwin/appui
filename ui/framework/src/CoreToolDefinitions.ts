/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

import {
  IModelApp, SelectionTool, FitViewTool, WindowAreaTool,
  PanViewTool, RotateViewTool, ViewToggleCameraTool, WalkViewTool, FlyViewTool, ZoomViewTool,
} from "@bentley/imodeljs-frontend";

// cSpell:ignore configurableui
import { ToolItemDef } from "./configurableui/Item";
import { ContentViewManager } from "./configurableui/ContentViewManager";
import { BaseItemState } from "./configurableui/ItemDefBase";
import { SyncUiEventId } from "./SyncUiEventDispatcher";
import { AnalysisAnimationTool } from "./tools/AnalysisAnimation";

/** Utility Class that provides definitions of tools provided by imodel.js core. These definitions can be used to populate the ui. */
export class CoreTools {
  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: "icon-fit-to-view",
      labelKey: "CoreTools:tools.View.Fit.flyover",
      tooltipKey: "CoreTools:tools.View.Fit.description",
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: "icon-window-area",
      labelKey: "CoreTools:tools.View.WindowArea.flyover",
      tooltipKey: "CoreTools:tools.View.WindowArea.description",
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: "icon-zoom",
      labelKey: "CoreTools:tools.View.Zoom.flyover",
      tooltipKey: "CoreTools:tools.View.Zoom.description",
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: "icon-hand-2",
      labelKey: "CoreTools:tools.View.Pan.flyover",
      tooltipKey: "CoreTools:tools.View.Pan.description",
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: "icon-rotate-left",
      labelKey: "CoreTools:tools.View.Rotate.flyover",
      tooltipKey: "CoreTools:tools.View.Rotate.description",
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: "icon-walk",
      labelKey: "CoreTools:tools.View.Walk.flyover",
      tooltipKey: "CoreTools:tools.View.Walk.description",
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: "icon-cursor",
      labelKey: "CoreTools:tools.Select.flyover",
      tooltipKey: "CoreTools:tools.Select.description",
      execute: () => { IModelApp.tools.run(SelectionTool.toolId); },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: "icon-camera",
      labelKey: "UiFramework:tools.View.ToggleCamera.flyover",
      tooltipKey: "UiFramework:tools.View.ToggleCamera.description",
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: "icon-airplane",
      labelKey: "UiFramework:tools.View.Fly.flyover",
      tooltipKey: "UiFramework:tools.View.Fly.description",
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get analysisAnimationCommand() {
    return new ToolItemDef({
      toolId: AnalysisAnimationTool.toolId,
      iconSpec: "icon-camera-animation",
      labelKey: "UiFramework:tools.AnalysisAnimation.flyover",
      tooltipKey: "UiFramework:tools.AnalysisAnimation.description",
      execute: () => { IModelApp.tools.run(AnalysisAnimationTool.toolId); },
      stateSyncIds: [SyncUiEventId.ActiveContentChanged],
      stateFunc: (currentState: Readonly<BaseItemState>): BaseItemState => {
        const returnState: BaseItemState = { ...currentState };
        const activeContentControl = ContentViewManager.getActiveContentControl();

        if (activeContentControl && activeContentControl.viewport && (undefined !== activeContentControl.viewport.view.AnalysisStyle))
          returnState.isVisible = true;
        else
          returnState.isVisible = false;
        return returnState;
      },
    });
  }
}
