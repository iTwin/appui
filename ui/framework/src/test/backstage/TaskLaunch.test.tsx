/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { mount, shallow } from "enzyme";
import { expect } from "chai";
import * as sinon from "sinon";
import {
  TaskLaunchBackstageItem,
  FrontstageManager,
  FrontstageActivatedEventArgs,
  ConfigurableUiManager,
  TaskPropsList,
  WorkflowPropsList,
  FrontstageProvider,
  Frontstage,
  FrontstageProps,
  BackstageItemState,
} from "../../ui-framework";
import TestUtils from "../TestUtils";
import { BackstageItem as NZ_BackstageItem } from "@bentley/ui-ninezone";
import { CoreTools } from "../../ui-framework/CoreToolDefinitions";
import { SyncUiEventDispatcher } from "../../ui-framework/syncui/SyncUiEventDispatcher";
import { WorkflowManager } from "../../ui-framework/workflow/Workflow";
import { Logger } from "@bentley/bentleyjs-core";

describe("Backstage", () => {
  const testEventId = "test-state-function-event";

  before(async () => {
    await TestUtils.initializeUiFramework();

    await FrontstageManager.setActiveFrontstageDef(undefined);
  });

  describe("<TaskLaunchBackstageItem />", () => {
    it("TaskLaunchBackstageItem should render & execute", async () => {
      class Frontstage1 extends FrontstageProvider {
        public get frontstage(): React.ReactElement<FrontstageProps> {
          return (
            <Frontstage
              id="Test1"
              defaultTool={CoreTools.selectElementCommand}
              defaultLayout="FourQuadrants"
              contentGroup="TestContentGroup1"
            />
          );
        }
      }
      const frontstageProvider = new Frontstage1();
      ConfigurableUiManager.addFrontstageProvider(frontstageProvider);

      const taskPropsList: TaskPropsList = {
        tasks: [
          {
            id: "Task1",
            primaryStageId: "Test1",
            iconSpec: "icon-placeholder",
            labelKey: "SampleApp:backstage.task1",
          },
        ],
      };

      ConfigurableUiManager.loadTasks(taskPropsList);

      // Test Workflows
      const workflowPropsList: WorkflowPropsList = {
        defaultWorkflowId: "default-workflow",
        workflows: [
          {
            id: "ExampleWorkflow",
            iconSpec: "icon-placeholder",
            labelKey: "SampleApp:Test.my-label",
            defaultTaskId: "task1",
            tasks: ["Task1"],
          },
        ],
      };

      let stateFuncRun = false;
      const stateFunc = (state: Readonly<BackstageItemState>): BackstageItemState => {
        stateFuncRun = true;
        return { ...state, isActive: true } as BackstageItemState;
      };

      ConfigurableUiManager.loadWorkflows(workflowPropsList);

      const spyMethod = sinon.stub();
      const remove = FrontstageManager.onFrontstageActivatedEvent.addListener((_args: FrontstageActivatedEventArgs) => spyMethod());
      const wrapper = mount(
        <TaskLaunchBackstageItem taskId="Task1" workflowId="ExampleWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder"
          isEnabled={true} isActive={false}
          stateSyncIds={[testEventId]} stateFunc={stateFunc} />,
      );
      const backstageItem = wrapper.find(NZ_BackstageItem);

      expect(stateFuncRun).to.be.false;
      SyncUiEventDispatcher.dispatchImmediateSyncUiEvent(testEventId);
      expect(stateFuncRun).to.be.true;

      backstageItem.find(".nz-backstage-item").simulate("click");
      await TestUtils.flushAsyncOperations();
      expect(spyMethod.calledOnce).to.be.true;
      remove();
      wrapper.unmount();
    });

    it("TaskLaunchBackstageItem should log error when invalid workflowId is provided", async () => {
      const spyMethod = sinon.spy(Logger, "logError");
      const wrapper = mount(
        <TaskLaunchBackstageItem taskId="Task1" workflowId="BadWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />,
      );
      const backstageItem = wrapper.find(NZ_BackstageItem);
      backstageItem.find(".nz-backstage-item").simulate("click");
      spyMethod.calledOnce.should.true;
      wrapper.unmount();
      (Logger.logError as any).restore();
    });

    it("TaskLaunchBackstageItem should log error when invalid taskId is provided", async () => {
      const spyMethod = sinon.spy(Logger, "logError");
      const wrapper = mount(
        <TaskLaunchBackstageItem taskId="BadTask" workflowId="ExampleWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />,
      );
      const backstageItem = wrapper.find(NZ_BackstageItem);
      backstageItem.find(".nz-backstage-item").simulate("click");
      spyMethod.calledOnce.should.true;
      wrapper.unmount();
      (Logger.logError as any).restore();
    });

    it("TaskLaunchBackstageItem renders correctly when inactive", () => {
      WorkflowManager.setActiveWorkflow(undefined);
      const wrapper = shallow(<TaskLaunchBackstageItem taskId="Task1" workflowId="ExampleWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />);
      wrapper.should.matchSnapshot();
      wrapper.unmount();
    });

    it("TaskLaunchBackstageItem renders correctly when active", async () => {
      const workflow = WorkflowManager.findWorkflow("ExampleWorkflow");
      expect(workflow).to.not.be.undefined;

      if (workflow) {
        const task1 = workflow.getTask("Task1");
        expect(task1).to.not.be.undefined;

        if (task1) {
          await WorkflowManager.setActiveWorkflowAndTask(workflow, task1);
          const wrapper = shallow(<TaskLaunchBackstageItem taskId="Task1" workflowId="ExampleWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" />);
          wrapper.should.matchSnapshot();
          wrapper.unmount();
        }
      }
    });

    it("TaskLaunchBackstageItem updates on property change", async () => {
      const wrapper = mount(<TaskLaunchBackstageItem taskId="Task1" workflowId="ExampleWorkflow" labelKey="UiFramework:tests.label" iconSpec="icon-placeholder" isEnabled={false} />);
      expect(wrapper.find("li.nz-disabled").length).to.eq(1);

      wrapper.setProps({ isEnabled: true });
      wrapper.update();
      expect(wrapper.find("li.nz-disabled").length).to.eq(0);

      wrapper.unmount();
    });

  });
});
