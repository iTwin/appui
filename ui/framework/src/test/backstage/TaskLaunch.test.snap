// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backstage <TaskLaunchBackstageItem /> TaskLaunchBackstageItem renders correctly when active 1`] = `
<BackstageItem
  icon={
    <Unknown
      iconSpec="icon-placeholder"
    />
  }
  isActive={true}
  isDisabled={false}
  key="ExampleWorkflow:Task1"
  onClick={[Function]}
>
  tests.label
</BackstageItem>
`;

exports[`Backstage <TaskLaunchBackstageItem /> TaskLaunchBackstageItem renders correctly when inactive 1`] = `
<BackstageItem
  icon={
    <Unknown
      iconSpec="icon-placeholder"
    />
  }
  isActive={false}
  isDisabled={false}
  key="ExampleWorkflow:Task1"
  onClick={[Function]}
>
  tests.label
</BackstageItem>
`;
